### ISSUE TITLE
aufs: trying to install perl-base in debian:jessie fails

### ISSUE THREAD

**Issue by jfrazelle** [docker/docker#19282]
```
Digest: sha256:f968f10b4b523737e253a97eac59b0d1420b5c19b69928d35801a6373ffe330e
Status: Downloaded newer image for debian:jessie
 ---> c648cd6a7396
Step 2 : RUN apt-get update && apt-get install -y apparmor bash-completion btrfs-tools build-essential curl ca-certificates debhelper dh-apparmor dh-systemd git libapparmor-dev libdevmapper-dev libltdl-dev libsqlite3-dev  libsystemd-journal-dev --no-install-recommends && rm -rf /var/lib/apt/lists/*

...

debconf: delaying package configuration, since apt-utils is not installed
Fetched 81.2 MB in 7min 37s (178 kB/s)
(Reading database ... 7531 files and directories currently installed.)
Preparing to unpack .../perl-base_5.20.2-3+deb8u2_amd64.deb ...
Unpacking perl-base (5.20.2-3+deb8u2) over (5.20.2-3+deb8u1) ...
dpkg: error processing archive /var/cache/apt/archives/perl-base_5.20.2-3+deb8u2_amd64.deb (--unpack):
 unable to make backup link of `./usr/bin/perl' before installing new version: No such file or directory
dpkg-deb: error: subprocess paste was killed by signal (Broken pipe)
Errors were encountered while processing:
 /var/cache/apt/archives/perl-base_5.20.2-3+deb8u2_amd64.deb
E: Sub-process /usr/bin/dpkg returned an error code (1)
The command '/bin/sh -c apt-get update && apt-get install -y apparmor bash-completion btrfs-tools build-essential curl ca-certificates debhelper dh-apparmor dh-systemd git libapparmor-dev libdevmapper-dev libltdl-dev libsqlite3-dev  libsystemd-journal-dev --no-install-recommends && rm -rf /var/lib/apt/lists/*' returned a non-zero code: 100

```

can reproduce when running `DOCKER_GRAPHDRIVER=aufs make deb`.

Seems to be aufs issue, dmesg has:

```
[17080.675373] aufs au_opts_verify:1597:docker[1175]: dirperm1 breaks the protection by the permission bits on the lower branch
[17080.742010] aufs au_opts_verify:1597:docker[1175]: dirperm1 breaks the protection by the permission bits on the lower branch
[17080.773853] aufs au_opts_verify:1597:docker[1175]: dirperm1 breaks the protection by the permission bits on the lower branch
[17080.812664] aufs au_opts_verify:1597:docker[1175]: dirperm1 breaks the protection by the permission bits on the lower branch
```

**Standard Info**
```
$ uname -a
Linux debian 4.3.3-fsociety #1 SMP Sat Dec 19 03:36:35 UTC 2015 x86_64 GNU/Linux

$ docker info
Containers: 4
Images: 83
Server Version: 1.10.0-dev
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 4.3.3-fsociety
Operating System: Debian GNU/Linux stretch/sid
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.359 GiB
Name: debian
ID: HID6:TAIG:SWCC:NFOG:KUDO:IOSY:LIY3:PJDB:CZJQ:IH6M:5HBP:6VUT
Debug mode (server): true
 File Descriptors: 41
 Goroutines: 55
 System Time: 2016-01-12T17:45:13.861380035-08:00
 EventsListeners: 0
 Init SHA1: 
 Init Path: /usr/bin/docker
 Docker Root Dir: /var/lib/docker

$ docker version
Client:
 Version:      1.10.0-dev
 API version:  1.22
 Go version:   go1.5.2
 Git commit:   09510c7
 Built:        Tue Jan 12 22:13:46 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.0-dev
 API version:  1.22
 Go version:   go1.5.2
 Git commit:   09510c7
 Built:        Tue Jan 12 22:13:46 2016
 OS/Arch:      linux/amd64
```

ping @tianon who has also seen it

**Comment by tianon** [comment]
I usually see something about noplink in dmesg when I hit the issue; will
see if I can get the exact log later this evening.

**Comment by tianon** [comment]
Simplest definite reproducer so far:
```console
$ docker run --rm debian@sha256:f968f10b4b523737e253a97eac59b0d1420b5c19b69928d35801a6373ffe330e sh -c 'apt-get update -qq && apt-get install -y perl-base'
Reading package lists...
Building dependency tree...
Reading state information...
Suggested packages:
  perl
The following packages will be upgraded:
  perl-base
1 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Need to get 1224 kB of archives.
After this operation, 455 kB of additional disk space will be used.
Get:1 http://security.debian.org/ jessie/updates/main perl-base amd64 5.20.2-3+deb8u2 [1224 kB]
debconf: delaying package configuration, since apt-utils is not installed
Fetched 1224 kB in 0s (1936 kB/s)
(Reading database ... 7531 files and directories currently installed.)
Preparing to unpack .../perl-base_5.20.2-3+deb8u2_amd64.deb ...
Unpacking perl-base (5.20.2-3+deb8u2) over (5.20.2-3+deb8u1) ...
dpkg: error processing archive /var/cache/apt/archives/perl-base_5.20.2-3+deb8u2_amd64.deb (--unpack):
 unable to make backup link of `./usr/bin/perl' before installing new version: No such file or directory
dpkg-deb: error: subprocess paste was killed by signal (Broken pipe)
Errors were encountered while processing:
 /var/cache/apt/archives/perl-base_5.20.2-3+deb8u2_amd64.deb
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

**Comment by tianon** [comment]
A little more interesting info:
```console
$ docker run -it --rm debian:jessie@sha256:f968f10b4b523737e253a97eac59b0d1420b5c19b69928d35801a6373ffe330e
root@dff82fec58bf:/# apt-get update
Get:1 http://security.debian.org jessie/updates InRelease [63.1 kB]
Ign http://httpredir.debian.org jessie InRelease          
Get:2 http://httpredir.debian.org jessie-updates InRelease [136 kB]
Get:3 http://httpredir.debian.org jessie Release.gpg [2373 B]
Get:4 http://httpredir.debian.org jessie Release [148 kB]         
Get:5 http://security.debian.org jessie/updates/main amd64 Packages [236 kB]
Get:6 http://httpredir.debian.org jessie-updates/main amd64 Packages [3619 B]
Get:7 http://httpredir.debian.org jessie/main amd64 Packages [9035 kB]
Fetched 9625 kB in 6s (1444 kB/s)                                              
Reading package lists... Done
aproot@dff82fec58bf:/# apt-get install perl-base
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Suggested packages:
  perl
The following packages will be upgraded:
  perl-base
1 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
Need to get 1224 kB of archives.
After this operation, 455 kB of additional disk space will be used.
Get:1 http://security.debian.org/ jessie/updates/main perl-base amd64 5.20.2-3+deb8u2 [1224 kB]
Fetched 1224 kB in 0s (1743 kB/s)    
debconf: delaying package configuration, since apt-utils is not installed
(Reading database ... 7531 files and directories currently installed.)
Preparing to unpack .../perl-base_5.20.2-3+deb8u2_amd64.deb ...
Unpacking perl-base (5.20.2-3+deb8u2) over (5.20.2-3+deb8u1) ...
dpkg: error processing archive /var/cache/apt/archives/perl-base_5.20.2-3+deb8u2_amd64.deb (--unpack):
 unable to make backup link of `./usr/bin/perl' before installing new version: No such file or directory
dpkg-deb: error: subprocess paste was killed by signal (Broken pipe)
E: Sub-process /usr/bin/dpkg returned an error code (1)
root@dff82fec58bf:/# mv /usr/bin/perl /usr/bin/perl.bak
mv: cannot move '/usr/bin/perl' to '/usr/bin/perl.bak': Input/output error
```

Which also gets me this in `dmesg`:
```
[102455.905063] aufs au_cpup_single:713:kworker/0:1[9101]: I/O Error, hi57937693(i156) exists on b0 but plink is disabled
```

But doing `mv /usr/bin/perl /usr/bin/perl.bak` before the stuff APT does has no effect.

**Comment by tianon** [comment]
(so there must be something else `dpkg` does before it tries to `mv /usr/bin/perl /usr/bin/perl.bak` that triggers the issue and makes it reproducible)

**Comment by tianon** [comment]
The following makes this a little easier to play with:
```Dockerfile
FROM debian:jessie@sha256:f968f10b4b523737e253a97eac59b0d1420b5c19b69928d35801a6373ffe330e

RUN apt-get update -qq && apt-get download perl-base

RUN dpkg --debug=12 -i perl-base_*.deb
```

(Adjusting `--debug` and/or skipping that line entirely and going interactive from the previous layer.)

```
Step 3 : RUN dpkg --debug=12 -i perl-base_*.deb
 ---> Running in b20daa584797
D000010: ensure_pathname_nonexisting '/var/lib/dpkg/tmp.ci'
(Reading database ... 7531 files and directories currently installed.)
Preparing to unpack perl-base_5.20.2-3+deb8u2_amd64.deb ...
D000002: maintscript_fallback nonexistent prerm '/var/lib/dpkg/info/perl-base.prerm'
Unpacking perl-base (5.20.2-3+deb8u2) over (5.20.2-3+deb8u1) ...
D000002: maintscript_new nonexistent preinst '/var/lib/dpkg/tmp.ci/preinst'
D000010: tarobject ti->name='.' mode=40755 owner=0:0 type=53(d) ti->linkname='' namenode='/.' flags=2 instead='<none>'
D000010: tarobject ti->name='./usr' mode=40755 owner=0:0 type=53(d) ti->linkname='' namenode='/usr' flags=2 instead='<none>'
D000010: tarobject ti->name='./usr/bin' mode=40755 owner=0:0 type=53(d) ti->linkname='' namenode='/usr/bin' flags=2 instead='<none>'
D000010: tarobject ti->name='./usr/bin/perl5.20.2' mode=100755 owner=0:0 type=48(-) ti->linkname='' namenode='/usr/bin/perl5.20.2' flags=2 instead='<none>'
D000010: ensure_pathname_nonexisting '/usr/bin/perl5.20.2.dpkg-new'
D000010: ensure_pathname_nonexisting '/usr/bin/perl5.20.2.dpkg-tmp'
D000010: tarobject ti->name='./usr/bin/perl' mode=100755 owner=0:0 type=49(h) ti->linkname='./usr/bin/perl5.20.2' namenode='/usr/bin/perl' flags=2 instead='<none>'
D000010: ensure_pathname_nonexisting '/usr/bin/perl.dpkg-new'
D000010: ensure_pathname_nonexisting '/usr/bin/perl.dpkg-tmp'
dpkg: error processing archive perl-base_5.20.2-3+deb8u2_amd64.deb (--install):
 unable to make backup link of `./usr/bin/perl' before installing new version: No such file or directory
D000010: cu_installnew '/usr/bin/perl' flags=2
D000010: cu_installnew '/usr/bin/perl5.20.2' flags=402
D000002: maintscript_new nonexistent postrm '/var/lib/dpkg/tmp.ci/postrm'
dpkg-deb: error: subprocess paste was killed by signal (Broken pipe)
D000002: vmaintscript_installed nonexistent postinst
D000010: ensure_pathname_nonexisting '/var/lib/dpkg/tmp.ci'
D000010: ensure_pathname_nonexisting running rm -rf '/var/lib/dpkg/tmp.ci'
D000010: ensure_pathname_nonexisting '/var/lib/dpkg/reassemble.deb'
Errors were encountered while processing:
 perl-base_5.20.2-3+deb8u2_amd64.deb
Removing intermediate container b20daa584797
```

**Comment by tianon** [comment]
Again with a more verbose debug value:
```
Step 3 : RUN dpkg --debug=3773 -i perl-base_*.deb
 ---> Running in 399de1639192
D000001: ensure_diversions: new, (re)loading
D000001: ensure_statoverrides: new, (re)loading
D000010: ensure_pathname_nonexisting '/var/lib/dpkg/tmp.ci'
(Reading database ... 7531 files and directories currently installed.)
Preparing to unpack perl-base_5.20.2-3+deb8u2_amd64.deb ...
D000001: process_archive oldversionstatus=installed
D000002: maintscript_fallback nonexistent prerm '/var/lib/dpkg/info/perl-base.prerm'
D000002: maintscript_new nonexistent preinst '/var/lib/dpkg/tmp.ci/preinst'
Unpacking perl-base (5.20.2-3+deb8u2) over (5.20.2-3+deb8u1) ...
D000010: tarobject ti->name='.' mode=40755 owner=0:0 type=53(d) ti->linkname='' namenode='/.' flags=2 instead='<none>'
D000100: setupvnamevbs main='/.' tmp='/..dpkg-tmp' new='/..dpkg-new'
D000100: tarobject already exists
D000100: tarobject directory exists
D000010: tarobject ti->name='./usr' mode=40755 owner=0:0 type=53(d) ti->linkname='' namenode='/usr' flags=2 instead='<none>'
D000100: setupvnamevbs main='/usr' tmp='/usr.dpkg-tmp' new='/usr.dpkg-new'
D000100: tarobject already exists
D000100: tarobject directory exists
D000010: tarobject ti->name='./usr/bin' mode=40755 owner=0:0 type=53(d) ti->linkname='' namenode='/usr/bin' flags=2 instead='<none>'
D000100: setupvnamevbs main='/usr/bin' tmp='/usr/bin.dpkg-tmp' new='/usr/bin.dpkg-new'
D000100: tarobject already exists
D000100: tarobject directory exists
D000010: tarobject ti->name='./usr/bin/perl5.20.2' mode=100755 owner=0:0 type=48(-) ti->linkname='' namenode='/usr/bin/perl5.20.2' flags=2 instead='<none>'
D000100: setupvnamevbs main='/usr/bin/perl5.20.2' tmp='/usr/bin/perl5.20.2.dpkg-tmp' new='/usr/bin/perl5.20.2.dpkg-new'
D000100: tarobject already exists
D000010: ensure_pathname_nonexisting '/usr/bin/perl5.20.2.dpkg-new'
D000010: ensure_pathname_nonexisting '/usr/bin/perl5.20.2.dpkg-tmp'
D000100: tarobject file open size=10416
D000100: tarobject file hash=93d68f39b1f58b5b3e9fcff0278f9257
D000100: tarobject nondirectory, 'link' backup
D000100: tarobject done and installation deferred
D000010: tarobject ti->name='./usr/bin/perl' mode=100755 owner=0:0 type=49(h) ti->linkname='./usr/bin/perl5.20.2' namenode='/usr/bin/perl' flags=2 instead='<none>'
D000100: setupvnamevbs main='/usr/bin/perl' tmp='/usr/bin/perl.dpkg-tmp' new='/usr/bin/perl.dpkg-new'
D000100: tarobject already exists
D000010: ensure_pathname_nonexisting '/usr/bin/perl.dpkg-new'
D000010: ensure_pathname_nonexisting '/usr/bin/perl.dpkg-tmp'
D000100: tarobject hardlink hash=93d68f39b1f58b5b3e9fcff0278f9257
D000100: tarobject nondirectory, 'link' backup
dpkg: error processing archive perl-base_5.20.2-3+deb8u2_amd64.deb (--install):
 unable to make backup link of `./usr/bin/perl' before installing new version: No such file or directory
D000010: cu_installnew '/usr/bin/perl' flags=2
D000100: setupvnamevbs main='/usr/bin/perl' tmp='/usr/bin/perl.dpkg-tmp' new='/usr/bin/perl.dpkg-new'
D000100: cu_installnew not restoring
D000100: secure_remove '/usr/bin/perl.dpkg-new' unlink OK
D000010: cu_installnew '/usr/bin/perl5.20.2' flags=402
D000100: setupvnamevbs main='/usr/bin/perl5.20.2' tmp='/usr/bin/perl5.20.2.dpkg-tmp' new='/usr/bin/perl5.20.2.dpkg-new'
D000100: cu_installnew restoring atomic
D000100: secure_remove '/usr/bin/perl5.20.2.dpkg-new' unlink OK
D000002: maintscript_new nonexistent postrm '/var/lib/dpkg/tmp.ci/postrm'
dpkg-deb: error: subprocess paste was killed by signal (Broken pipe)
D000002: vmaintscript_installed nonexistent postinst
D000010: ensure_pathname_nonexisting '/var/lib/dpkg/tmp.ci'
D000010: ensure_pathname_nonexisting running rm -rf '/var/lib/dpkg/tmp.ci'
D000010: ensure_pathname_nonexisting '/var/lib/dpkg/reassemble.deb'
Errors were encountered while processing:
 perl-base_5.20.2-3+deb8u2_amd64.deb
Removing intermediate container 399de1639192
```

**Comment by tianon** [comment]
For reference, here's my kernel info:
```
Linux nameless 4.1.15-gentoo #2 SMP Thu Dec 17 15:07:48 PST 2015 x86_64 AMD FX(tm)-8120 Eight-Core Processor AuthenticAMD GNU/Linux
```
(so it's not something specific to 4.3 -- might still be AUFS version; @justincormack poked me to say he can't reproduce on newer AUFS + 4.1 so I'm going to test a more recent b2d build now that it's got latest AUFS and see if I can't reproduce there)

**Comment by tianon** [comment]
Ok, can't reproduce on my b2d qemu:
```
Boot2Docker version 1.9.1, build master : 608e3e3 - Mon Jan 11 19:18:57 UTC 2016
Linux boot2docker 4.1.15-boot2docker #1 SMP Mon Jan 11 18:01:38 UTC 2016 x86_64 GNU/Linux
AUFS v4.1.13+-20160111
```

**Comment by tianon** [comment]
So whatever this issue is, it's fixed in newer versions of AUFS, which means that the fix for #18180 should cover this bug too (and Ubuntu's planning to apply it too: https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1533043).

**Comment by tianon** [comment]
Hmm, this doesn't seem to be necessarily conclusively fixed with just newer AUFS -- Gentoo has AUFS 20160118 now, but after rebooting with the updated module I still have this particular issue. :disappointed:

**Comment by tianon** [comment]
I wonder if it's possible that something between 20160111 and 20160118 broke it again? (https://github.com/sfjro/aufs4-standalone/compare/0e5c702a803ebfd9a38e5037e830da07c44c0af6...7254f7820ca90483896d2796b90b2a3028cdb0dc)

**Comment by tianon** [comment]
Ah, looks like Gentoo does have 20160111, so I'm testing that next. :+1:

**Comment by tianon** [comment]
Interesting, can still reproduce on 20160111 + 4.1.15 here.  Not sure what else to try to narrow this down. :cry:

**Comment by AkihiroSuda** [comment]
Cannot reproduce on kernel 4.2.0-23-generic (Ubuntu 15.10).
I think this is unrelated to #18180

**Comment by AkihiroSuda** [comment]
After updating kernel to 4.2.0-25 and docker (from 1.6.2) to 1.10.0-rc1, I can hit the bug on my Ubuntu machine.
I'll look into this.

**Comment by AkihiroSuda** [comment]
Possibly related post (AUFS mepis live system, posted very recently: Jan 19): http://forum.mepiscommunity.org/viewtopic.php?f=92&t=39329

**Comment by AkihiroSuda** [comment]
https://github.com/docker/docker/commit/ef05b83417e32f269daf798307adcf07fe6ef13f (merged since v1.10.0-rc1) introduced `noplink` mount option.
@tonistiigi Do you think this is a bug of AUFS or an expected AUFS behavior?

**Comment by justincormack** [comment]
I also got this only after upgrading to 1.10.0-rc1, and only noticed it with debian:unstable so I thought for a while they had broken debian but now realise it is this issue. Linux 4.1.15, aufs4.1.13+ 0e5c702a803ebfd9a38e5037e830da07c44c0af6

**Comment by duane-unitive** [comment]
Is there a fix for this problem? I have upgraded to 1.10.2 and I cannot install perl-base. I get the same error message.

**Comment by justincormack** [comment]
@duane-unitive can you give more information about your kernel kersion and aufs version and exactly which sequence of commands fails? I have no issues any more, just retested the ones in this thread.

**Comment by duane-unitive** [comment]
Apologies, I updated docker, but we are also using dinghy which I failed to upgrade properly. Upon upgrading dinghy as well, the build is working properly.

Thanks!
